apps/backend/src/mutations/FapMutations.ts

```diff
@@ -243,62 +244,108 @@ export default class FapMutations {
   @Authorized([Roles.USER_OFFICER])
   async assignProposalsToFapUsingCallInstrument(
     agent: UserWithRole | null,
-    args: AssignProposalsToFapUsingCallInstrumentArgs
+    args: AssignProposalsToFapsUsingCallInstrumentArgs
   ): Promise<boolean | Rejection> {
     return this.assignProposalsToFapUsingCallInstrumentInternal(agent, args);
   }
 
   async assignProposalsToFapUsingCallInstrumentInternal(
     agent: UserWithRole | null,
-    args: AssignProposalsToFapUsingCallInstrumentArgs
+    args: AssignProposalsToFapsUsingCallInstrumentArgs
   ): Promise<boolean | Rejection> {
     const proposals = await this.proposalDataSource.getProposalsByPks(
       args.proposalPks
     );
 
     const callHasInstruments =
-      await this.callDataSource.getCallHasInstrumentsByInstrumentId(
-        args.instrumentId
+      await this.callDataSource.getCallHasInstrumentsByInstrumentIds(
+        args.instrumentIds
       );
 
     const callIds = [...new Set(proposals.map((proposal) => proposal.callId))];
     for (const callId of callIds) {
-      const callHasInstrument = callHasInstruments.find(
-        (callHasInstrument) => callHasInstrument.callId === callId
-      );
-      if (callHasInstrument && callHasInstrument.fapId) {
-        await this.assignProposalsToFapInternal(agent, {
-          proposals: proposals
-            .filter((proposal) => proposal.callId === callId)
-            .map((proposal) => ({
-              ...proposal,
-              callId: callId,
-            })),
-          fapId: callHasInstrument.fapId,
-          fapInstrumentId: callHasInstrument.instrumentId,
-        });
+      if (callHasInstruments.length) {
+        const fapInstruments = callHasInstruments
+          .filter((callHasInstrument) => callHasInstrument.fapId)
+          .map((callHasInstrument) => ({
+            fapId: callHasInstrument.fapId,
+            instrumentId: callHasInstrument.instrumentId,
+          }));
+        if (fapInstruments.length) {
+          await this.assignProposalsToFapsInternal(agent, {
```

Comment:
Would it be more efficient to move the `for (const callId of callIds) {` loop to only include this call? It seems unnecessary to be making the same `callHasInstruments` call every time round the loop.